### ISSUE TITLE
6. Machine Learning

### ISSUE THREAD

**Issue by p-jiho** [p-jiho/natural-language#6]
#### 1) 머신러닝이 아닌 접근 방법의 한계
- 이미지 인식과 같은 경우는 기존 프로그래밍 방식으로는 특징을 잡기가 힘들다.
- 여러 규칙을 정하고 특징을 잡아도 한계가 있어 사람이 규칙을 정하는 것이 아닌 머신러닝으로 해결이 가능하다.

#### 2) 머신러닝 방식
- 주어진 데이터를 통해 기계가 스스로 학습하며 규칙성을 찾는다.

<br></br>

#### 1) 머신 러닝 모델의 평가
- 데이터는 훈련용, 테스트용, 검증용으로 분리한다.
- 검증용 데이터 : 모델 성능을 조정하기 위한 것이다. 과적합 판단이나 하이퍼파라미터 조정을 위한 것이다.
- 하이퍼파라미터 : 사용자가 직접 지정해주는 변수. 모델의 성능을 개선하며 계속 튜닝을 한다.
- 매개변수 : 모델이 학습하는 과정에서 얻어져 계속 변하는 수

#### 2) Classification과 Regression
1. Binary Classification
- 두 개의 선택지 중 하나의 답을 선택하는 경우
2. Multi-class Classification
- 세 개 이상의 선택지 중 답을 선택하는 경우
3. Regression
- 위 두 분류 문제처럼 정해진 답이 있는 것이 아닌 연속된 결과를 가지는 문제

#### 3) 지도 학습과 비지도 학습
1. Supervised Learning
: 레이블이라는 정답이 있는 데이터를 학습시키는 것이다. 자연어 처리는 대부분 지도학습이다.
2. Unsupervised Learning
: 레이블이 없이 학습을 하는 것이다. LSA나 LDA가 속한다.
3. Self-Supervised Learning, SSL
: 레이블 없는 데이터가 주어졌을 때, 스스로 레이브를 만들어 학습하는 경우이다. Word2Vec와 같은 워드 임베딩 알고리즘이나, BERT가 속한다.

#### 4) Sample과 Feature
- 머신 러닝의 많은 모델들은 행렬 연산을 한다.
- Sample : 데이터 행렬 중 행
- Feature : 데이터 행렬 중 열. y를 예측하기 위한 x의 특성

#### 5) Confusion Matrix
- 정확도 = (맞춘 문제수)/(전체 문제수)
- 혼동행렬 : 맞고 틀린 결과 뿐만 아니라 세부적인 내용을 알려준다.
- **TP** : 실제 True, 예측 True, **FP** : 실제 False, 예측 True, **FN** : 실제 True, 예측 False, **TN** : 실제 False, 예측 False
1. Precision
- 정밀도 = TP/(TP+FP) : 즉, True라고 예측한 것 중 실제 True인 비율
2. Recall
- 재현율 = TP/(TP+FN) : 즉, 실제 True인 것 중 True라고 맞게 예측한 비율
3. Accurancy
- 정확도 = (TP+TN)/(TP+FN+FP+TN) : 전체 중에서 맞게 예측한 비율
- 정확도는 전체 데이터에 비해 중요한 데이터가 너무 적은 비율이라면 유용한 결과를 얻을 수 없다. => F1-Score 사용

#### 6) Overfitting과 Underfitting
- 과적합 : 데이터를 과도하게 학습한 경우이다. 예측할 데이터는 많은데 소수의 데이터인 훈련용 데이터에만 과하게 맞춰있는 모델은 좋은 모델이 아니다.
    -> 테스트 데이터로 평가를 해보면 과적합의 경우 오차가 엄청 크다.
- epoch : 데이터에 대한 훈련 반복 횟수이다. 에포크가 크면 과적합이 발생한다.
- 과소적합 : 훈련이 부족하여 성능이 발전할 수 있음에도 멈추는 것이다. 훈련 데이터에 대한 정확도도 낮다.
- 학습 과정
    - 데이터를 훈련용 데이터, 검증용 데이터, 테스트용 데이터로 나눈다. 훈련용 데이터를 좀 크게 잡는다.
    - 훈련용 데이터로 학습하고 검증용 데이터로 평가하는 과정을 반복하다가 검증용 데이터의 오차가 커지면 과적합이므로 멈춘다.
    - 완성된 모델을 테스트용 데이터로 평가를 한다.

<br></br>

#### 1) Cost function : MSE
- 목적함수, 비용함수, 손실함수 : 실제값과 예측값에 대한 오차. 셋이 같은 의미는 아님
- MSE : 예측값의 오차를 줄이기 위해 사용. 오차 제곱합의 평균

#### 2) Optimizer : Gradient Descent
- 경사하강법 : 비용 함수를 미분해 0인 곳을 찾아가는 방법
- 비용을 최소화하는 w 업데이트 : 학습률을 이용해 w를 조정한다.
- 학습률 : w가 변경될 때 어느정도로 크게 어느 방향으로 변경할지를 결정한다. 0과 1사이의 값을 갖는다.
    -> 지나치게 크면 w가 발산을 하고, 지나치게 작으면 학습의 속도가 느려진다.

<br></br>

#### 1) 자동 미분
- tensorflow의 자동 미분 기능 : https://www.tensorflow.org/guide/autodiff?hl=ko

#### 2) 자동 미분을 이용한 선형 회귀 구현
1. `tf.Variable` 함수를 이용해 변수 w와 b를 선언, 가설 함수를 선언
2. `tf.square`함수로 실제값과 예측값의 차이의 제곱을 구한 후 `tf.reduce_mean`을 이용해 평균을 구하는 함수 생성
3. x, y 데이터를 선언한 후 `tf.optimizers.SGD`함수를 이용해 학습률을 지정
4. `tf.GradientTape` 함수를 이용해 MSE와 예측값 계산
5. `tape.gradient` 함수를 이용해 미분 값 계산 후 `optimizer.apply_gradients` 함수를 이용해 파라미터를 업데이터 한다.
6. for 문을 이용해 여러번 반복 하여 `w.numpy()`, `b.numpy()`, `cost` 변수를 이용해 값을 추출

#### 3) 케라스로 구현하는 선형 회귀
```
from tensorflow.keras.models ipmort Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras import optimizers
```
1. 모델 생성 : `model = Sequential()`
2. 필요한 정보 추가 : `model.add(Dense(1, input_dim=1, activation="linear"))`, 첫번째 인자 : y의 차원, `input_dim`인자 : x의 차원, `activation` : 어떤 회귀모형인지
3. 학습률 지정 : `sgd = optimizers.SGD(lr=0.01)`, 학습률은 0.01로 지정됌
4. 모델 컴파일 : `model.compile(optimizer=sgd, loss="mse", metrics= ["mse"])`, MSE를 사용
5. 모델 적합 : `model.fit(x,y, epochs=300)`. x,y는 데이터, `epochs` : 반복 횟수
6. 예측값 확인 : `model.predict(x)`

<br></br>

#### 1) Binary Classification
- 이진 분류에서 선형회귀는 적절하지 않다.
- Sigmoid function : 0과 1사이의 값을 s 형으로 가지는 함수

#### 2) Sigmoid function
```
def sigmoid(x):
   return 1/(1+np.exp(-x))
```
- w 값이 커질수록 경사도가 커진다.
- b 값이 커질수록 1로 수렴을 한다.

#### 3) Cost function
- 로지스틱 회귀는  MSE를 사용하면 전체 함수의 최솟값이 아닌 특정한 구역에 있는 최솟값을 최솟값으로 착각할 수 있다.
- 목적함수 : 가중치를 최소화하는 함수
- Cross Entropy : 로지스틱 회귀에서 사용하는 비용함수. 크로스 엔트로피의 평균을 구해 사용

<br></br>

#### 1) 케라스로 구현하는 로지스틱 회귀
- 선형회귀와 하는 방식은 같지만 세부사항이 다르다.
1. 모델 생성 : `model=Sequential()`
2. 필요한 정보 추가 : `model.add(Dense(1, input_dim=1,activaation="sigmoid"))`, 시그모이드 함수를 사용하므로 `activation` 인자를 `sigmoid`로 지정한다.
3. 학습률 지정 : `sgd = optimizers.SGD(lr=0.01)`
4. 모델 컴파일 : `model.compile(optimizer=sgd, loss = "binary_crossentropy",metrics=["binary_accuracy"])`, 크로스 엔트로피를 사용하고 이진 분류 문제이다.
5. 모델 적합 : `model.fit(x,y,epochs=200)

<br></br>

#### 1) 다중 선형 회귀
- 이 전 회귀식들과 달리 입력 변수가 2개 이상인 경우이다.
- `model.add` 함수에서 `input_dim` 인자가 입력 변수의 차원이므로 차원에 맞게 값을 입력해주면 다중회귀를 할 수 있다.

#### 2) 다중 로지스틱 회귀
- 이 전 로지스틱 회귀식에서 `input_dim` 인자의 수만 변화시킨 것이다.

<br></br>

#### 1) 벡터와 행렬과 텐서
- 벡터 : 크기와 방향을 가진 것. 1차원 배열
- 행렬 : 행과 열을 가지는 2차원 구조
- 텐서 : 3차원 이상

#### 2) Tensor
- Numpy의 `ndim` : 텐서의 차원
- Numpy의 `shape` : 텐서의 크기
1. 0차원 텐서(스칼라)
- 차원 : 0, 크기 : 없음
2. 1차원 텐서(벡터)
- 차원 : 1, 크기 : (n,)
- 주의 : 벡터의 크기를 차원이라고 하는데 벡터의 차원은 벡터 안의 요소의 갯수이고 텐서의 차원은 벡터의 경우 1차원이다.
3. 2차원 텐서(행렬)
- 차원 : 2, 크기 : (n,m)
- 텐서의 크기는 n개의 데이터 안에 각각 m개의 요소가 들어있는 것이다.
4. 3차원 텐서(다차원 배열)
- 차원 : 3, 크기 : (n,m,p)
- 3D 텐서 : 샘플의 갯수(samples, batch_size), 시퀀스의 길이(timesteps), 벡터의 차원(word_dim)으로 구성되어있다.
- 자연어 처리와 3D 텐서와의 관계?
    - 단어의 벡터화 : 원-핫 인코딩, 워드 임베딩
    - sample : 문서의 갯수
    - timesteps : 문서 안에 단어의 갯수
    - word_dim : 벡터화 된 단어들의 총 갯수
5. 그 이상의 텐서
- (n-1)차원 텐서들을 배열로 합해 n차원 텐서를 만든다.
6. 케라스에서의 텐서
- `input_shape` : batch_size를 제외한 차원을 지정
- `batch_input_shape` : 배치 크기까지 지정하는 인자

#### 3) 벡터와 행렬의 연산
1. 벡터의 내적과 행렬의 곱셈
- `np.dot` : 벡터의 내적
- `np.matmul` : 행렬의 곱셈

#### 4) 다중 선형 회귀 행렬 연산으로 이해하기
- 독립 변수가 2개 이상일 경우 : 입력 벡터와 가중치 벡터의 내적으로 다중 선형 회귀 표현이 가능하다.
- 샘플의 갯수가 많은 경우 : 입력 행렬과 가중치 벡터의 곱으로 표현이 가능하다.

<br></br>

#### 1) Multi-class Classification
- 소프트맥스 함수 : 3 개 이상으로 분류하는 다중 클래스 분류에서 사용한다. 도합 1의 확률로 계산을 해서 확률이 가장 높은 선택지를 선택한다.

#### 2) Softmax function
1. 차원 축소 : 가중치 곱을 이용해 차원의 수를 조정한다.
2. 오차 계산 : 확률로 나오는 예측값과 원-핫 벡터로 나오는 실제값의 차이를 계산한다. 예를 들어 0.2, 0.3,0.5 가 예측값이고 0 0 1이 실제 값이면 둘의 오차를 크로스 엔트로피 함수로 계산을 한다.

#### 3) 원-핫 벡터의 무작위성
- 정수 인코딩의 문제점 : 클래스를 분류하는 문제에서는 순서 정보가 없는 경우가 많다. 그러므로 각 클래스 간의 오차가 동일해야하는데 정수 인코딩을 시행하면 클래스 간의 차이가 생기게 된다.
- 원-핫 인코딩의 장점 : 클래스 표현 방법이 무작위성을 가져 클래스 사이에 순서나 차이가 발생하지 않는다.

#### 4) Cost function
1. 크로스 엔트로피 함수
: 정확하게 예측한 경우 크로스 엔트로피 함수의 값은 0이 된다. 예측에서 멀어지면 값이 커진다. -> 최종 합의 값이 최소화되도록 학습해야한다.
2. 이진 분류에서의 크로스 엔트로피 함수
: 로지스틱 회귀에서의 크로스 엔트로피 함수와 한 눈에 보기엔 차이가 있지만 본질적으로는 같은 함수이다.
  소프트맥스 회귀에서의 크로스 엔트로피 함수에서 k=2라고 하면 식이 같아진다.

#### 5) 인공 신경망 프로그램
: 소프트맥스 회귀는 인공 신경망과 중 하나라고 볼 수 있다.

<br></br>

## 10. 소프트맥스 회귀 실습
1. 데이터 확인 : seaborn의 pairplot으로 산점도를 그려보고, barplot을 그려보는 등 데이터를 여러 방면으로 확인해본다.
2. 데이터 수치화 : `replace` 함수를 이용해 레이블을 정수인코딩한다.
3. 레이블 분리 : 훈련데이터와 테스트 데이터를 `train_test_split` 함수로 나눈다.
4. 원-핫 인코딩 : 정수 인코딩했던 열을 `to_categorical`함수를 이용해 원-핫 인코딩을 해주며 순서를 없앤다.
5. 모델 생성 후 필요한 정보 추가 : `Sequential()`함수를 이용해 모델 생성 후 `model.add`함수를 이용해 필요한 정보를 넣는다. 소프트 맥스 함수를 사용하면 `activation="softmax"`로 인자를 넣어준다.
6. 모델 컴파일 : `model.compile` 함수를 사용한다. 다중 클래스 분류 문제 에서는 `loss = "categorical_crossentropy"` 인자를 넣어준다. 
7. 모델 적합 : `model.fit`함수를 이용해 모델 적합을 한다.
8. 정확도 확인 : `evaluate`함수를 이용해 정확도 출력

---

- 참고
>> 딥 러닝을 이용한 자연어 처리 입문, 유원준, 안상준
      * 위 글 내용은 이 책의 내용을 공부하여 정리한 것입니다.
